ember-cli/2017-03/march-31.md

```diff
@@ -0,0 +1,127 @@
+  - composability of blueprints
+  - blueprint migrations/transactional?
+- Background Service
+- Slow Commands
+  - `generate`
+  - `help`
+- Ember CLI/Broccoli Errors
+- Global add-ons
+- `lib-cli`
+
+## Discussion Notes
+
+### Glimmer Pipeline
+
+Extract `lib/broccoli/ember-app.js` into a separate npm package.
+
+### Tree Shaking
+
+- given list of file entries + module information + roots, generate a retained graph
+- keep module loose w/o concat until the very end
+- generate dep graph at the very end of the build process
+- have a hook that will give you access to dep graph w/o being able to modify it
+
+### `app.import` from `node_modules`
+
+- support amd, cjs or es6 w/o watching
+- everything that's imported is considered dynamically linked
+- should care about single file (entry point) make sure we can tell the difference between globals/modules for tree shaking (globals and module are in different buckets)
+- need to be able to "blacklist" modules to resolve them during runtime
+  - possibly make it automatic, we *should* have info already or do a dep walk on app.import path
+  - should *not* require for users to pass into list of modules when using `app.import`
+- address differences between `import`s (documentation or otherwise)
+  - `app.import` is project relative
+  - `this.import` is relative to whatever `this` is
+- path *has* to start with `vendor`, `node_modules`
+
+### Blrpnt extraction
+
+- unfork
+  - port new test changes from ember-cli
+  - make sure we haven't lost anything added in ember-cli/blueprint.js
+- load blprnt in ember-cli
+  - bring into ember-cli as drop in replacement
+- decouple pods/classic structures information from blprint/blueprint.js
+  - create new structures using new primitive to manage this logic in ember-cli
+- create module unification structure to provide blueprint support
+  - use glimmer resolver to take advantage of rules to generate file paths
+
+### Background Service
+
+**idea 1**: Convert bin/cli into an actor system to provide parallelism  (actor === task)
+
+**idea 2**: Daemonize the process
+
+- removing add-ons (packages.json changes) needs to restart daemon
+- daemon per application
+- abstraction at app layer
+
+*Conclusion*: Since the gain of this refactor doesn't seem high enough
+at this point we will not move forward with those plans.
+
+### Slow Commands
+
+#### `generate`:
+
+- idea 1: cache the project property that gets passed in, which is probably the biggest slowdown
+- idea 2: based on ember-cli version, add-on commands will use a shim for `includedCommands`
+  to export a hash with keys being the command name, and value being a hash including the path
+  to the module to be required, as well as command options and any other relevant info
+  - the primary issue comes from requires at the top of the command
+- deprecate command name property, in favor of the key
+
+#### `help`
+
+- upon add-on install, identify and write an add-on's available blueprint g -h json
```

Comment:
are you talking about `require()`ing the commands eagerly? we discussed whether it would be possible to only declare the commands the addon has without actually loading the command code already. unfortunately the `aliases` property of the commands seem to make this a little harder.